File: android/app/src/main/java/com/retainic/app/ui/StatsScreen.kt
package com.retainic.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.BarChart
import androidx.compose.material.icons.filled.Psychology
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.Icon
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.foundation.Canvas
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.retainic.app.R
import com.retainic.app.data.AuthService
import com.retainic.app.data.GlossaryEntry
import com.retainic.app.data.GlossaryRepository
import com.retainic.app.data.VocabRepository
import com.retainic.app.data.VocabWord
import kotlinx.coroutines.async
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

private val AspectColors = listOf(Color(0xFF3B82F6), Color(0xFF22C55E), Color(0xFFF59E0B))

/**
 * Glossary practice records the same daily tallies as list practice, so the
 * totals cover both a user's words and their glossary terms.
 */
/**
 * How fast one kind of item is being memorized: how many of them are done, over
 * the days since the first one was added.
 */
private class Pace(val count: Int, val memorized: Int, createdDates: List<Date>) {
    val startDate: Date? = createdDates.minByOrNull { it.time }
    val activeDays: Int = if (startDate != null) {
        maxOf(1, VocabWord.daysBetween(VocabWord.startOfDay(startDate), VocabWord.startOfDay(Date())) + 1)
    } else 1
    val perDay: Double = memorized.toDouble() / activeDays
    val perWeek: Double = perDay * 7
    val perMonth: Double = perDay * (365.25 / 12)
}

private class LearningStats(allWords: List<VocabWord>, allEntries: List<GlossaryEntry> = emptyList()) {
    val totalWords = allWords.size + allEntries.size
    val totalMemorized = allWords.count { it.isRemembered } + allEntries.count { it.isRemembered }

    // Words and terms are paced separately: each counts from the day its own
    // first one was added, so a glossary started last week isn't judged against
    // months of vocabulary.
    val words = Pace(allWords.size, allWords.count { it.isRemembered }, allWords.mapNotNull { it.createdAt })
    val terms = Pace(allEntries.size, allEntries.count { it.isRemembered }, allEntries.mapNotNull { it.createdAt })
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun StatsScreen(auth: AuthService, modifier: Modifier = Modifier) {
    var stats by remember { mutableStateOf<LearningStats?>(null) }
    var today by remember { mutableStateOf(mapOf("word" to 0, "translation" to 0, "pronunciation" to 0)) }
    var week by remember { mutableStateOf<List<Triple<Date, String, Int>>>(emptyList()) }
    var glossary by remember { mutableStateOf(GlossaryStats()) }
    var glossaryWeek by remember { mutableStateOf<List<Triple<Date, String, Int>>>(emptyList()) }
    var isLoading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(auth.uid) {
        val uid = auth.uid ?: return@LaunchedEffect
        isLoading = true
        try {
            // Everything this screen needs, fetched at once: the words, the
            // glossary terms and the week's tallies don't depend on each other,
            // so waiting for each in turn only added up round trips.
            val allDeferred = async { VocabRepository.fetchAllWords(uid) }
            val entriesDeferred = async { GlossaryRepository.fetchAllEntries(uid) }
            val dailyDeferred = async {
                runCatching { VocabRepository.fetchDailyStats(uid, 7) }.getOrDefault(emptyList())
            }
            val all = allDeferred.await()
            val allEntries = entriesDeferred.await()
            stats = LearningStats(all, allEntries)
            today = countTodayRemembered(all)
            glossary = glossaryStats(allEntries)
            val daily = dailyDeferred.await()
            week = buildWeekPoints(daily, today)
            glossaryWeek = buildGlossaryWeekPoints(daily, glossary)
        } catch (e: Exception) {
            error = e.localizedMessage
        } finally {
            isLoading = false
        }
    }

    Scaffold(
        modifier = modifier,
        topBar = { TopAppBar(title = { Text(stringResource(R.string.statistics)) }) },
    ) { inner ->
        Box(Modifier.padding(inner).fillMaxSize()) {
            val s = stats
            when {
                isLoading && s == null -> LoadingView(stringResource(R.string.loading))
                s == null || s.totalWords == 0 -> EmptyState(
                    icon = Icons.Filled.BarChart,
                    title = stringResource(R.string.no_stats_yet),
                    description = stringResource(R.string.no_stats_desc),
                )
                else -> StatsContent(s, today, week, glossary, glossaryWeek)
            }
        }
    }

    ErrorDialog(error) { error = null }
}

@Composable
private fun StatsContent(
    stats: LearningStats,
    today: Map<String, Int>,
    week: List<Triple<Date, String, Int>>,
    glossary: GlossaryStats,
    glossaryWeek: List<Triple<Date, String, Int>>,
) {
    val aspectKeys = listOf("word", "translation", "pronunciation")
    val aspectLabels = listOf(stringResource(R.string.word), stringResource(R.string.translation), stringResource(R.string.pronunciation))

    Column(
        Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(24.dp),
    ) {
        // Total memorized card
        Column(
            Modifier.fillMaxWidth()
                .background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(20.dp))
                .padding(vertical = 28.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Icon(Icons.Filled.Psychology, contentDescription = null, modifier = Modifier.size(44.dp),
                tint = MaterialTheme.colorScheme.primary)
            Text("${stats.totalMemorized}", fontSize = 52.sp, fontWeight = FontWeight.Bold)
            Text(stringResource(R.string.words_and_terms_memorized), style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
            Text(stringResource(R.string.out_of_n_total, stats.totalWords),
                style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }

        // Words practice today (bar chart)
        Text(stringResource(R.string.words_practice_today), style = MaterialTheme.typography.titleMedium)
        BarChart(aspectKeys.mapIndexed { i, k -> Triple(aspectLabels[i], today[k] ?: 0, AspectColors[i]) })
        Text(stringResource(R.string.n_cards_practised, aspectKeys.sumOf { today[it] ?: 0 }),
            style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center, modifier = Modifier.fillMaxWidth())

        // Words this week (line chart)
        Text(stringResource(R.string.words_this_week), style = MaterialTheme.typography.titleMedium)
        WeekLineChart(week, aspectKeys, aspectLabels, AspectColors)

        // Glossaries get their own charts: what was practised today, and the
        // week behind it.
        if (glossary.total > 0) {
            Text(stringResource(R.string.glossary_practice_today), style = MaterialTheme.typography.titleMedium)
            BarChart(listOf(
                Triple(stringResource(R.string.term), glossary.termsToday, AspectColors[0]),
                Triple(stringResource(R.string.definition), glossary.definitionsToday, AspectColors[1]),
            ))
            Text(stringResource(R.string.n_cards_practised, glossary.termsToday + glossary.definitionsToday),
                style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center, modifier = Modifier.fillMaxWidth())

            // The glossary week is drawn from tallies glossary practice writes
            // for itself, so it only fills in from the day that started.
            Text(stringResource(R.string.glossary_this_week), style = MaterialTheme.typography.titleMedium)
            WeekLineChart(
                glossaryWeek,
                listOf("glossaryTerm", "glossaryDefinition"),
                listOf(stringResource(R.string.term), stringResource(R.string.definition)),
                AspectColors,
            )
        }

        // Average pace, words and terms apart
        if (stats.words.count > 0) PaceSection(stringResource(R.string.words_average_pace), stats.words)
        if (stats.terms.count > 0) PaceSection(stringResource(R.string.terms_average_pace), stats.terms)
    }
}

/** One kind's pace, with the stretch of learning it was measured over. */
@Composable
private fun PaceSection(title: String, pace: Pace) {
    Text(title, style = MaterialTheme.typography.titleMedium)
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        PaceCard(stringResource(R.string.per_day), pace.perDay, Modifier.weight(1f))
        PaceCard(stringResource(R.string.per_week), pace.perWeek, Modifier.weight(1f))
        PaceCard(stringResource(R.string.per_month), pace.perMonth, Modifier.weight(1f))
    }
    pace.startDate?.let { start ->
        val since = DateFormat.getDateInstance(DateFormat.MEDIUM, Locale.getDefault()).format(start)
        Text(stringResource(R.string.based_on_days, pace.activeDays, since),
            style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center, modifier = Modifier.fillMaxWidth())
    }
}

@Composable
private fun PaceCard(title: String, value: Double, modifier: Modifier) {
    Column(
        modifier.background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(16.dp))
            .padding(vertical = 18.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        val formatted = if (value < 10) String.format(Locale.getDefault(), "%.1f", value)
        else String.format(Locale.getDefault(), "%.0f", value)
        Text(formatted, style = MaterialTheme.typography.titleLarge, fontWeight = FontWeight.Bold)
        Text(title, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
    }
}

/** Every chart is drawn in a box this tall, so the dashboard lines up. */
private val ChartHeight = 220.dp

/** The bar area of a bar chart, below the value labels and above the names. */
private val BarPlotHeight = 150.dp

/**
 * The values to rule a chart at: the finest round step that keeps it to five
 * lines or fewer, so the gaps are easy to read off. A chart whose tallest value
 * is 3 gets lines at 1, 2 and 3; one reaching 50 gets 10, 20, 30, 40, 50.
 */
private fun guideValues(max: Int): List<Int> {
    val step = listOf(1, 2, 5, 10, 20, 25, 50, 100, 200, 250, 500, 1000)
        .firstOrNull { (max + it - 1) / it <= 5 } ?: 1000
    return generateSequence(step) { it + step }.takeWhile { it <= max }.toList()
}

/**
 * Dashed guide lines behind a chart, so a bar or a point can be read against a
 * value instead of eyeballed. [plotHeight] is the height the data is drawn in,
 * measured up from the bottom of this canvas; null means the whole canvas.
 */
@Composable
private fun GuideLines(max: Int, plotHeight: Dp? = null, modifier: Modifier = Modifier) {
    val color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.3f)
    Canvas(modifier) {
        val plot = plotHeight?.toPx() ?: size.height
        val dash = PathEffect.dashPathEffect(floatArrayOf(8f, 8f))
        for (value in guideValues(max)) {
            val y = size.height - plot * (value.toFloat() / max)
            drawLine(color, Offset(0f, y), Offset(size.width, y), strokeWidth = 2f, pathEffect = dash)
        }
    }
}

@Composable
private fun BarChart(bars: List<Triple<String, Int, Color>>) {
    val max = (bars.maxOfOrNull { it.second } ?: 0).coerceAtLeast(1)
    Column(Modifier.fillMaxWidth().height(ChartHeight).padding(vertical = 8.dp)) {
        // The bars and their names are separate rows, so the bottom of this box
        // is the line the bars stand on — the one the guide lines measure from.
        Box(Modifier.fillMaxWidth().weight(1f)) {
            GuideLines(max, BarPlotHeight, Modifier.matchParentSize())
            // Each bar takes an equal share of the width, and each name below
            // takes the same share — so they stay centred on each other, and a
            // long name like "Pronunciation" has the whole share to sit in
            // rather than a box narrower than the word.
            Row(Modifier.fillMaxSize(), verticalAlignment = Alignment.Bottom) {
                bars.forEach { (_, count, color) ->
                    Column(
                        Modifier.weight(1f),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.Bottom,
                    ) {
                        Text("$count", style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant)
                        Box(
                            Modifier.width(48.dp)
                                .height((BarPlotHeight * (count.toFloat() / max)).coerceAtLeast(2.dp))
                                .background(color, RoundedCornerShape(6.dp)),
                        )
                    }
                }
            }
        }
        Row(Modifier.fillMaxWidth()) {
            bars.forEach { (label, _, _) ->
                Text(
                    label,
                    style = MaterialTheme.typography.bodySmall,
                    maxLines = 1,
                    softWrap = false,
                    overflow = TextOverflow.Ellipsis,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.weight(1f).padding(top = 4.dp, start = 2.dp, end = 2.dp),
                )
            }
        }
    }
}

@Composable
private fun WeekLineChart(
    points: List<Triple<Date, String, Int>>,
    aspectKeys: List<String>,
    aspectLabels: List<String>,
    colors: List<Color>,
) {
    // Group by aspect key -> ordered 7 daily values.
    val days = points.map { it.first }.distinct().sortedBy { it.time }
    val max = (points.maxOfOrNull { it.third } ?: 0).coerceAtLeast(1)
    val weekdayFmt = remember { SimpleDateFormat("EEE", Locale.getDefault()) }

    val guideColor = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.3f)

    Column(Modifier.fillMaxWidth()) {
        Canvas(Modifier.fillMaxWidth().height(ChartHeight).padding(vertical = 8.dp)) {
            val n = days.size
            if (n < 2) return@Canvas
            val stepX = size.width / (n - 1)
            fun y(v: Int) = size.height - (v.toFloat() / max) * size.height
            // Ruled first, so the week's lines are drawn over the guides.
            val dash = PathEffect.dashPathEffect(floatArrayOf(8f, 8f))
            for (value in guideValues(max)) {
                drawLine(guideColor, Offset(0f, y(value)), Offset(size.width, y(value)),
                    strokeWidth = 2f, pathEffect = dash)
            }
            aspectKeys.forEachIndexed { ai, key ->
                val series = days.map { d ->
                    points.firstOrNull { it.first == d && it.second == key }?.third ?: 0
                }
                for (i in 0 until n - 1) {
                    drawLine(
                        color = colors[ai],
                        start = Offset(i * stepX, y(series[i])),
                        end = Offset((i + 1) * stepX, y(series[i + 1])),
                        strokeWidth = 4f,
                    )
                }
                series.forEachIndexed { i, v ->
                    drawCircle(colors[ai], radius = 5f, center = Offset(i * stepX, y(v)))
                }
            }
        }
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            days.forEach { d ->
                Text(weekdayFmt.format(d), style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
        }
        // Legend
        Row(Modifier.fillMaxWidth().padding(top = 8.dp), horizontalArrangement = Arrangement.spacedBy(16.dp)) {
            aspectLabels.forEachIndexed { i, label ->
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Box(Modifier.size(10.dp).background(colors[i], RoundedCornerShape(50)))
                    Text(" $label", style = MaterialTheme.typography.labelSmall)
                }
            }
        }
    }
}

/**
 * What a user's glossary practice looks like today, for the charts glossaries
 * get to themselves. [total] is how many terms there are at all, which decides
 * whether those charts are worth showing.
 */
data class GlossaryStats(
    val total: Int = 0,
    val termsToday: Int = 0,
    val definitionsToday: Int = 0,
)

private fun glossaryStats(entries: List<GlossaryEntry>): GlossaryStats {
    val now = Date()
    fun isToday(d: Date?) = d != null && VocabWord.isSameDay(d, now)
    var termsToday = 0; var definitionsToday = 0
    for (e in entries) {
        if (isToday(e.lastTermRemembered)) termsToday++
        // Every definition is a card of its own, so each one recalled today
        // counts on its own.
        definitionsToday += e.definitionList.count { isToday(it.lastRemembered) }
    }
    return GlossaryStats(entries.size, termsToday, definitionsToday)
}

/**
 * Counts words whose per-aspect last-remembered date is today. Glossary terms
 * are counted apart, in [GlossaryStats], so the two sets of charts don't count
 * the same practice twice.
 */
private fun countTodayRemembered(words: List<VocabWord>): Map<String, Int> {
    val now = Date()
    fun isToday(d: Date?) = d != null && VocabWord.isSameDay(d, now)
    var word = 0; var translation = 0; var pronunciation = 0
    for (w in words) {
        if (isToday(w.lastWordRemembered)) word++
        if (isToday(w.lastTranslationRemembered)) translation++
        if (isToday(w.lastPronounciationRemembered)) pronunciation++
    }
    return mapOf("word" to word, "translation" to translation, "pronunciation" to pronunciation)
}

/**
 * Builds 7 days of (date, glossary aspect key, count) from the tallies glossary
 * practice writes for itself. Today comes from the terms, as it does in the
 * combined chart.
 */
private fun buildGlossaryWeekPoints(
    daily: List<com.retainic.app.data.DailyStat>,
    glossary: GlossaryStats,
): List<Triple<Date, String, Int>> {
    val cal = Calendar.getInstance()
    val todayStart = VocabWord.startOfDay(Date())
    val byKey = daily.associateBy { it.date }
    val result = mutableListOf<Triple<Date, String, Int>>()
    for (offset in 6 downTo 0) {
        cal.time = todayStart
        cal.add(Calendar.DAY_OF_YEAR, -offset)
        val day = cal.time
        val stat = byKey[VocabRepository.dayKey(day)]
        result.add(Triple(day, "glossaryTerm",
            if (offset == 0) glossary.termsToday else stat?.glossaryTerm ?: 0))
        result.add(Triple(day, "glossaryDefinition",
            if (offset == 0) glossary.definitionsToday else stat?.glossaryDefinition ?: 0))
    }
    return result
}

/** Builds 7 days of (date, aspectKey, count): today from the words, earlier from the log. */
private fun buildWeekPoints(daily: List<com.retainic.app.data.DailyStat>, today: Map<String, Int>): List<Triple<Date, String, Int>> {
    val cal = Calendar.getInstance()
    val todayStart = VocabWord.startOfDay(Date())
    val byKey = daily.associateBy { it.date }
    val aspectKeys = listOf("word", "translation", "pronunciation")
    val result = mutableListOf<Triple<Date, String, Int>>()
    for (offset in 6 downTo 0) {
        cal.time = todayStart
        cal.add(Calendar.DAY_OF_YEAR, -offset)
        val day = cal.time
        val stat = byKey[VocabRepository.dayKey(day)]
        for (key in aspectKeys) {
            // A day's shared tally covers both kinds of practice, so the word
            // series is what's left once the glossary's own tally is taken out.
            // Days logged before glossaries tallied separately have nothing to
            // take out, and read as before.
            val value = if (offset == 0) today[key] ?: 0 else when (key) {
                "word" -> maxOf(0, (stat?.word ?: 0) - (stat?.glossaryTerm ?: 0))
                "translation" -> maxOf(0, (stat?.translation ?: 0) - (stat?.glossaryDefinition ?: 0))
                "pronunciation" -> stat?.pronunciation ?: 0
                else -> 0
            }
            result.add(Triple(day, key, value))
        }
    }
    return result
}

File: web/js/screens/stats-screen.js
//
//  Statistics aggregation and charts.
//  Retainic Web
//

import { el, clear, svgEl } from "../dom.js";
import * as i18n from "../i18n.js";
import { t, tn, tf } from "../i18n.js";
import * as Repo from "../repository.js";
import * as M from "../models.js";
import * as G from "../glossary.js";
import { authState } from "../auth.js";
import { navBar, spinner, emptyState, errorState, icon } from "../ui.js";

export async function StatsScreen(content) {
  content.appendChild(navBar(t("Statistics"), {}));
  const body = el(".scroll");
  content.appendChild(body);
  body.appendChild(spinner(t("Loading…")));

  let words = [];
  let entries = [];
  let dailyStats = [];
  try {
    const lists = await Repo.fetchLists(authState.uid);
    for (const l of lists) words = words.concat(await Repo.fetchWords(authState.uid, l.id));
    // Glossary practice writes the same daily tallies as list practice, so the
    // totals and today's counts have to include glossary terms as well.
    const glossaries = await Repo.fetchGlossaries(authState.uid);
    for (const g of glossaries) entries = entries.concat(await Repo.fetchEntries(authState.uid, g.id));
    dailyStats = await Repo.fetchDailyStats(authState.uid, 7).catch(() => []);
  } catch (e) { clear(body); body.appendChild(errorState(e)); return; }

  clear(body);
  if (words.length + entries.length === 0) {
    body.appendChild(emptyState(icon("bar_chart", 46), t("No Stats Yet"),
      t("Add words and practice them. Once you've memorized some, your progress shows up here.")));
    return;
  }

  // Aggregate stats (LearningStats). A word counts as memorized only once it is
  // fully remembered: 8× word, 10× translation, 7× pronunciation (remember_final);
  // a glossary term once the term and each of its definitions have had their
  // five recalls.
  const totalWords = words.length + entries.length;
  const totalMemorized = words.filter(M.isRemembered).length + entries.filter(G.isRemembered).length;
  const dates = [...words, ...entries].map((w) => w.createdAt).filter(Boolean);
  const start = dates.length ? new Date(Math.min(...dates.map((d) => +d))) : null;
  let activeDays = 1;
  if (start) {
    const s = new Date(start); s.setHours(0, 0, 0, 0);
    const n = new Date(); n.setHours(0, 0, 0, 0);
    activeDays = Math.max(1, Math.round((n - s) / 86400000) + 1);
  }
  const perDay = totalMemorized / activeDays;
  const perWeek = perDay * 7;
  const perMonth = perDay * (365.25 / 12);

  // Today remembered (derived from words and glossary terms). A term's two
  // methods line up with a word's first two: recalling the term itself, and
  // recalling what it means.
  const today = { word: 0, translation: 0, pronunciation: 0 };
  for (const w of words) {
    if (M.isToday(w.lastWordRemembered)) today.word += 1;
    if (M.isToday(w.lastTranslationRemembered)) today.translation += 1;
    if (M.isToday(w.lastPronounciationRemembered)) today.pronunciation += 1;
  }
  for (const e of entries) {
    if (M.isToday(e.lastTermRemembered)) today.word += 1;
    // Every definition is a card of its own, so each one recalled today counts
    // — the same way the daily tallies were written during practice.
    today.translation += G.definitions(e).filter((d) => M.isToday(d.lastRemembered)).length;
  }

  // Glossary progress, for the charts glossaries get to themselves. A term is
  // memorized once its term side and every one of its definitions have had
  // their five recalls; anything with at least one recall behind it is under
  // way, and the rest hasn't been started.
  const glossaryToday = { term: 0, definition: 0 };
  const glossaryProgress = { memorized: 0, started: 0, untouched: 0 };
  for (const e of entries) {
    if (M.isToday(e.lastTermRemembered)) glossaryToday.term += 1;
    const defs = G.definitions(e);
    glossaryToday.definition += defs.filter((d) => M.isToday(d.lastRemembered)).length;
    const recalls = (e.timesTermCorrect ?? 0) + defs.reduce((sum, d) => sum + d.timesCorrect, 0);
    if (G.isRemembered(e)) glossaryProgress.memorized += 1;
    else if (recalls > 0) glossaryProgress.started += 1;
    else glossaryProgress.untouched += 1;
  }

  // The glossary curve reads the fields glossary practice tallies for itself.
  const glossaryKeys = ["glossaryTerm", "glossaryDefinition"];
  const glossaryLabel = (k) => (k === "glossaryTerm" ? t("Term") : t("Definition"));
  const glossaryColors = { glossaryTerm: "#2f6bff", glossaryDefinition: "#1fb56a" };
  const glossaryWeekToday = {
    glossaryTerm: glossaryToday.term,
    glossaryDefinition: glossaryToday.definition,
  };

  const aspectKeys = ["word", "translation", "pronunciation"];
  const aspectLabel = (k) => k === "word" ? t("Word") : k === "translation" ? t("Translation") : t("Pronunciation");
  const colors = { word: "#2f6bff", translation: "#1fb56a", pronunciation: "#ff8a1f" };

  body.appendChild(el(".stats", {},
    // total card
    el(".stat-total", {},
      el(".big-icon", {}, icon("psychology", 44)),
      el(".stat-number", {}, `${totalMemorized}`),
      el(".stat-caption", {}, t("words and terms memorized")),
      el(".stat-subcaption", {}, tf("out of %lld total", totalWords)),
    ),
    el(".stat-block", {},
      el("h3", {}, t("Words today")),
      barChart(aspectKeys.map((k) => ({ label: aspectLabel(k), value: today[k], color: colors[k] }))),
    ),
    el(".stat-block", {},
      el("h3", {}, t("Words this week")),
      weekChart(dailyStats, today, aspectKeys, aspectLabel, colors),
    ),
    // Glossaries get their own two charts: how far the terms have come, and
    // what was practised today. Only shown once there are terms to chart.
    entries.length ? el(".stat-block", {},
      el("h3", {}, t("Glossary terms")),
      barChart([
        { label: t("Memorized"), value: glossaryProgress.memorized, color: colors.word },
        { label: t("In progress"), value: glossaryProgress.started, color: colors.translation },
        { label: t("Not started"), value: glossaryProgress.untouched, color: colors.pronunciation },
      ]),
      el("p.caption.center", {}, tf("%lld of %lld terms memorized", glossaryProgress.memorized, entries.length)),
    ) : null,
    entries.length ? el(".stat-block", {},
      el("h3", {}, t("Glossary practice today")),
      barChart([
        { label: t("Term"), value: glossaryToday.term, color: colors.word },
        { label: t("Definition"), value: glossaryToday.definition, color: colors.translation },
      ]),
      el("p.caption.center", {}, tn("%lld cards practised", glossaryToday.term + glossaryToday.definition)),
    ) : null,
    // The glossary week is drawn from tallies glossary practice writes for
    // itself, so it only fills in from the day that started — today always
    // comes from the terms themselves, as it does in the combined chart.
    entries.length ? el(".stat-block", {},
      el("h3", {}, t("Glossary this week")),
      weekChart(dailyStats, glossaryWeekToday, glossaryKeys, glossaryLabel, glossaryColors),
    ) : null,
    el(".stat-block", {},
      el("h3", {}, t("Average pace")),
      el(".pace-row", {},
        paceCard(t("Per day"), perDay),
        paceCard(t("Per week"), perWeek),
        paceCard(t("Per month"), perMonth),
      ),
    ),
    start ? el("p.caption.center", {},
      tf("Based on %lld days of learning since %@.", activeDays, start.toLocaleDateString(i18n.preferredLanguage()))) : null,
  ));

  function paceCard(title, value) {
    const text = value < 10 ? value.toFixed(1) : `${Math.round(value)}`;
    return el(".pace-card", {}, el(".pace-value", {}, text), el(".pace-title", {}, title));
  }
}

function barChart(bars) {
  const W = 320, H = 200, pad = 28;
  const max = Math.max(1, ...bars.map((b) => b.value));
  const bw = (W - pad * 2) / bars.length;
  const svg = svgEl("svg", { viewBox: `0 0 ${W} ${H}`, class: "chart", preserveAspectRatio: "xMidYMid meet" });
  bars.forEach((b, i) => {
    const h = (b.value / max) * (H - pad * 2);
    const x = pad + i * bw + bw * 0.2;
    const w = bw * 0.6;
    const y = H - pad - h;
    svg.appendChild(svgEl("rect", { x, y, width: w, height: h, rx: 6, fill: b.color }));
    svg.appendChild(svgEl("text", { x: x + w / 2, y: y - 6, "text-anchor": "middle", class: "chart-val" }, document.createTextNode(`${b.value}`)));
    const label = svgEl("text", { x: x + w / 2, y: H - 8, "text-anchor": "middle", class: "chart-lbl" });
    label.appendChild(document.createTextNode(b.label));
    svg.appendChild(label);
  });
  return svg;
}

function weekChart(dailyStats, today, aspectKeys, aspectLabel, colors) {
  const W = 340, H = 220, padX = 24, padY = 28;
  const byDate = {};
  for (const s of dailyStats) byDate[s.date] = s;
  const days = [];
  const now = new Date(); now.setHours(0, 0, 0, 0);
  for (let off = 6; off >= 0; off--) {
    const d = new Date(now); d.setDate(d.getDate() - off);
    const key = Repo.dayKey(d);
    const stat = byDate[key];
    const vals = {};
    for (const k of aspectKeys) {
      vals[k] = off === 0 ? (today[k] || 0) : (stat ? (stat[k] || 0) : 0);
    }
    days.push({ date: d, vals });
  }
  const max = Math.max(1, ...days.flatMap((d) => aspectKeys.map((k) => d.vals[k])));
  const innerW = W - padX * 2, innerH = H - padY * 2;
  const x = (i) => padX + (i / 6) * innerW;
  const y = (v) => padY + innerH - (v / max) * innerH;
  const svg = svgEl("svg", { viewBox: `0 0 ${W} ${H}`, class: "chart", preserveAspectRatio: "xMidYMid meet" });
  // gridlines + weekday labels
  days.forEach((d, i) => {
    svg.appendChild(svgEl("line", { x1: x(i), y1: padY, x2: x(i), y2: padY + innerH, stroke: "#e6e6ec", "stroke-width": 1, "stroke-dasharray": "4 4" }));
    const lbl = svgEl("text", { x: x(i), y: H - 8, "text-anchor": "middle", class: "chart-lbl" });
    lbl.appendChild(document.createTextNode(d.date.toLocaleDateString(i18n.preferredLanguage(), { weekday: "narrow" })));
    svg.appendChild(lbl);
  });
  for (const k of aspectKeys) {
    const pts = days.map((d, i) => `${x(i)},${y(d.vals[k])}`).join(" ");
    svg.appendChild(svgEl("polyline", { points: pts, fill: "none", stroke: colors[k], "stroke-width": 2.5, "stroke-linejoin": "round", "stroke-linecap": "round" }));
    days.forEach((d, i) => svg.appendChild(svgEl("circle", { cx: x(i), cy: y(d.vals[k]), r: 3, fill: colors[k] })));
  }
  // legend
  const legend = el(".chart-legend", {}, ...aspectKeys.map((k) =>
    el(".legend-item", {}, el(".legend-dot", { style: `background:${colors[k]}` }), aspectLabel(k))));
  return el(".chart-wrap", {}, svg, legend);
}
